src/doc/trpl/the-stack-and-the-heap.md

```diff
@@ -1,3 +1,526 @@
+| 0x01       | i    | 0x40000000 |
+| 0x00       | h    | 3          |
+
+Regardless, you may see diagrams in other examples that use hexadecimal. We’ll
+just stick to decimal, since we’re already halfway through the example.
+
+At the end of `bar()`, it calls `baz()`:
+
+| Address      | Name | Value        |
++--------------+------+--------------+
+|  2 ^ 30      |      | 20           |
+| (2 ^ 30) - 1 |      | 5            |
+| ...          | ...  | ...          |
+| 12           | g    | 100          |
+| 11           | f    | 4            |
+| 10           | e    | 4            |
+| 9            | d    | (2 ^ 30) - 1 |
+| 8            | c    | 5            |
+| 7            | b    | 4            |
+| 6            | a    | 0            |
+| 5            | z    | 4            |
+| 4            | y    | 10           |
+| 3            | x    | 0            |
+| 2            | j    | 0            |
+| 1            | i    | 2 ^ 30       |
+| 0            | h    | 3            |
+
+With this, we’re at our deepest point! Whew! Congrats for following along this
+far.
+
+After `baz()` is over, we get rid of `f` and `g`:
+
+| Address      | Name | Value        |
++--------------+------+--------------+
+|  2 ^ 30      |      | 20           |
+| (2 ^ 30) - 1 |      | 5            |
+| ...          | ...  | ...          |
+| 10           | e    | 4            |
+| 9            | d    | (2 ^ 30) - 1 |
+| 8            | c    | 5            |
+| 7            | b    | 4            |
+| 6            | a    | 0            |
+| 5            | z    | 4            |
+| 4            | y    | 10           |
+| 3            | x    | 0            |
+| 2            | j    | 0            |
+| 1            | i    | 2 ^ 30       |
+| 0            | h    | 3            |
+
+Next, we return from `bar()`. `d` in this case is a `Box<T>`, so it also frees
+what it points to: `(2 ^ 30) - 1`.
+
+| Address | Name | Value  |
++---------+------+--------+
+|  2 ^ 30 |      | 20     |
+| ...     | ...  | ...    |
+| 5       | z    | 4      |
+| 4       | y    | 10     |
+| 3       | x    | 0      |
+| 2       | j    | 0      |
+| 1       | i    | 2 ^ 30 |
+| 0       | h    | 3      |
+
+And after that, `foo()` returns:
+
+| Address | Name | Value  |
++--------------+------+---+
+|  2 ^ 30 |      | 20     |
+| ...     | ...  | ...    |
+| 2       | j    | 0      |
+| 1       | i    | 2 ^ 30 |
+| 0       | h    | 3      |
+
+And then, finally, `main()`, which cleans the rest up. When `i` is `Drop`ped,
+it will clean up the last of the heap too.
+
+# What do other languages do?
+
+Most languages with a garbage collector heap-allocate by default. This means
+that every value is boxed. There are a number of reasons why this is done, but
+they’re out of scope for this tutorial. There are some possible optimizations
+that don’t make it true 100% of the time, too. Rather than relying on the stack
+and `Drop` to clean up memory, the garbage collector deals with the heap
+instead.
```

Comment:
The reason I had the parenthetical: "(at a cost: ...)" was in anticipation of the response, "If heap-allocation is strictly more general, then why is this document making such a fuss about stack-allocation."  But maybe its worth dedicating a bit more text to this point; this is after all the "the stack and the heap" chapter, so shouldn't it address "which to use: stack versus heap" ?

The big picture answer to "stack versus heap" that I see is: the LIFO model of the stack is simpler, at a *fundamental* level.  This has two big impacts:
 1. Runtime Efficiency.
  * Managing the memory for the stack is trivial: The machine just increments or decrements a single value, the so-called "stack pointer".
  * Managing memory for the heap is non-trivial: since heap-allocated memory is freed at arbitrary points, and each block of heap-allocated memory can be of arbitrary size, the memory manager must generally work much harder to identify memory for reuse.  (I would then refer the reader to [Wilson et al., 1995], ([Wilson alt. link]) and [Mem Mgmt Wikipedia])
[Wilson et al., 1995]: http://dl.acm.org/citation.cfm?id=664690
[Wilson alt. link]: http://www.cs.northwestern.edu/~pdinda/icsclass/doc/dsa.pdf
[Mem Mgmt Wikipedia]: http://en.wikipedia.org/wiki/Memory_management
 2. Semantic impact (this is far more important, IMO): Stack-allocation impacts the Rust language itself, and thus the developer's mental model.
  * The LIFO semantics is what drives how the Rust language handles automatic memory management.  Even the de-allocation of a heap-allocated box can be driven by the stack-based LIFO semantics, as discussed throughout this chapter.